packages/grid/_modules_/grid/hooks/features/filter/gridFilterSelector.ts

```diff
@@ -12,15 +12,41 @@ export const visibleGridRowsStateSelector = (state: GridState) => state.visibleR
 export const visibleSortedGridRowsSelector = createSelector<
   GridState,
   VisibleGridRowsState,
-  GridRowModel[],
-  GridRowModel[]
->(
-  visibleGridRowsStateSelector,
-  sortedGridRowsSelector,
-  (visibleRowsState, sortedRows: GridRowModel[]) => {
-    return [...sortedRows].filter((row) => visibleRowsState.visibleRowsLookup[row.id] !== false);
-  },
-);
+  Map<GridRowId, GridRowModel>,
+  Map<GridRowId, GridRowModel>
+>(visibleGridRowsStateSelector, sortedGridRowsSelector, (visibleRowsState, sortedRows) => {
+  const map = new Map();
+  sortedRows.forEach((row, id) => {
```

Comment:
As visible rows might be smaller than sorted rows, it's better to loop through visible rows I think
How about?
```tsx
const union = new Map(
  [...visibleMap.entries()]
    .filter(entry=> entry[1])
    .map(entry=>[entry[0], sortMap.get(entry[0])])
)
```

Reply:
To loop first the visible rows we have to convert it to a map or it won't work with numeric ids. It's better to deal with this in https://github.com/mui-org/material-ui-x/issues/874 which will be another PR.